rfcs/2021-07-19-8216-multiple-pipelines.md

```diff
@@ -0,0 +1,258 @@
+inputs = ["generate"]
+```
+
+But it becomes possible with
+
+```toml
+# /etc/vector/first.toml
+[sources.generate-first]
+...
+[resources.my-pipeline]
+generate = ["generate-first"]
+# /etc/vector/second.toml
+[sources.generate-second]
+...
+[resources.my-pipeline]
+generate = ["generate-second"]
+# /etc/vector/pipelines/my-pipeline.yml
+[transforms.something]
+type = "remap"
+inputs = ["generate"]
+```
+
+The pipeline resource would have the following structure
+
+```rust
+struct PipelineResource {
+  filename: String,
+  aliases: Map<String, Set<String>>,
+  config: PipelineConfig,
+}
+```
+
+The structure of the pipeline's configuration file is as follows
+
+```toml
+# optional
+id = "load-balancer"
+
+[transforms.first]
+inputs = ["source"]
+...
+
+[transforms.second]
+inputs = ["first"]
+outputs = ["exposed-sink"]
+...
+
+# this will error
+[transforms.third]
+inputs = ["other-pipeline.bar"]
+outputs = ["exposed-sink"]
+...
+```
+
+In order to allow the user to specify the targeted sinks by a pipeline, the `forward` component creates a forwarding to the real sinks from the configuration files.
+It also allows the user to use is own aliases and just refer the external sinks in a single place.
+
+Now, if we look deeper at the configuration building process, the configuration compiler will require the pipelines in order to build the [configuration](https://github.com/timberio/vector/blob/v0.15.0/src/config/builder.rs#L71).
+
+To do so, we'll need to implement a `PipelineConfigBuilder` with the following structures
+
+```rust
+struct PipelineTransformOuter {
+    #[serde(flatten)]
+    content: TransformOuter,
+    outputs: Set<String>,
+}
+struct PipelineConfigBuilder {
+    pub id: String,
+    pub transforms: Map<String, PipelineTransformOuter>,
+}
+```
+
+then we'll update [the `compile` function](https://github.com/timberio/vector/blob/v0.15.0/src/config/compiler.rs#L4), in order to build a `Config` containing the required pipeline components, the compiler will load the pipeline's configurations, load the transforms and substitute the aliases.
+
+The components coming from the pipeline would be cloned inside the final `Config`, in the `transforms` `IndexMap` and the `outputs` from the pipeline components will be added to the referring `Sink` input field.
+
+### Observing pipelines
+
+Users should be able to observe and monitor individual pipelines.
+This means relevant metrics coming from the `internal_metrics` source must contain a `pipeline_id` tag referring to the pipeline `id`.
+
+In Vector, the `Task` structure is what emits the events for `internal_metrics`.
+After [build the different pieces of the topology](https://github.com/timberio/vector/blob/v0.15.0/src/topology/builder.rs#L106), we've to update the [`Task::new`](https://github.com/timberio/vector/blob/v0.15.0/src/topology/builder.rs#L163) in order to accept an `Option<PipelineId>` so that when it emits the metrics events it can provide the information about the pipeline.
```

Comment:
This is hard to understand. The grammar is off. I tried fixing it myself but wasn't sure what you meant.